src/main/scala/com/fulcrumgenomics/bam/EstimateRnaSeqInsertSize.scala

```diff
@@ -0,0 +1,267 @@
+          val rec5Prime   = if (rec.getReadNegativeStrandFlag) rec.getAlignmentEnd else rec.getAlignmentStart
+          val mate5Prime  = if (rec.getMateNegativeStrandFlag) mateAlignmentEnd    else rec.getMateAlignmentStart
+
+          // calculate the distance in transcript space
+          val lower5Prime = Math.min(rec5Prime, mate5Prime)
+          val upper5Prime = Math.max(rec5Prime, mate5Prime)
+          val transcriptInsertSize = transcript.filter { exon => CoordMath.overlaps(lower5Prime, upper5Prime, exon.start, exon.end) }
+            .map { exon =>
+              CoordMath.getOverlap(lower5Prime, upper5Prime, exon.start, exon.end)
+            }
+            .sum
+
+          // Make sure that the insert size is the same as any previous result
+          insertSize match {
+            case None        => insertSize = Some(transcriptInsertSize)
+            case Some(isize) => sameInsertSize = isize == transcriptInsertSize
+          }
+        }
+      }
+
+      if (sameInsertSize) {
+        insertSize
+      }
+      else {
+        None
+      }
+    }
+    else {
+      None
+    }
+  }
+
+  /** Gets the number of read bases that overlap the exons. */
+  private def numReadBasesOverlappingTranscript(alignmentBlocks: List[AlignmentBlock], transcript: Transcript): Int = {
+    val blocks  = alignmentBlocks.iterator.bufferBetter
+    val exons   = transcript.iterator.bufferBetter
+    var overlap = 0
+
+    while (blocks.hasNext && exons.hasNext) {
+      val block               = blocks.head
+      val exon                = exons.head
+      val blockReferenceStart = block.getReferenceStart
+      val blockReferenceEnd   = blockReferenceStart + block.getLength
+      val exonReferenceStart  = exon.start
+      val exonReferenceEnd    = exon.end
+
+      // Calculate the number of read bases in alignment block overlap the exon
+      if (CoordMath.overlaps(blockReferenceStart, blockReferenceEnd, exonReferenceStart, exonReferenceEnd)) {
+        overlap += CoordMath.getOverlap(blockReferenceStart, blockReferenceEnd, exonReferenceStart, exonReferenceEnd)
+      }
+
+      // Consume either the next block or next exon, based on which has the earlier end position
+      if (blockReferenceEnd < exonReferenceEnd) {
+        blocks.next()
+      }
+      else {
+        exons.next()
+      }
+    }
+
+    overlap
+  }
+}
+
+object EstimateRnaSeqInsertSize {
+
+  val RnaSeqInsertSizeMetricExtension = ".rna_seq_insert_size.txt"
+
+  private def filters(minimumMappingQuality: Int, includeDuplicates: Boolean) = List(
+    new MateMappedFilter, // also filters out unpaired reads
+    new FailsVendorReadQualityFilter,
+    new AlignedFilter(true),
+    new SecondaryOrSupplementaryFilter,
+    new MappingQualityFilter(minimumMappingQuality),
+    new DuplicatesFilter(includeDuplicates=includeDuplicates),
+    new FirstOfPairOnlyFilter,
+    new DifferentReferenceIndexFilter
+  )
+
+  trait SingleEndSamRecordFilter extends SamRecordFilter {
```

Comment:
If you wanted to make this more compact I think you could do something like this:
```scala
object SamRecordFilters {
  private trait SeFilter extends SamRecordFilter { override def filterOut(r1: SAMRecord, r2: SAMRecord) = filterOut(r1) || filterOut(r2) } 
  def filter(f: SAMRecord => Boolean): SamRecordFilter = new SeFilter { 
    override def filterOut(r: SAMRecord): Boolean = f(r)
  }

  def ReadPairedFilter = filter(!_.getReadPairedFlat)
  def MateMappedFilter = filter(r => !r.getReadPairedFlag || r.getMateUnmappedFlag)
  ...
}
```